### PR TITLE
Ucc integration

### PR DESCRIPTION
**Pull request by kaiyingshan**

### CODE REVIEW

**Comment by nirandaperera** [comment]
@kaiyingshan I'm getting the following seg fault. I can't figure out where and why it is coming from. I'm using latest UCC master branch. 
```
(cylon_dev) niranda@aurora-r10:~/git/cylon/build$ ./bin/ucc_allgather_example 
[1656170916.449172] [aurora-r10:49843:0]          ucc_cl.c:57   UCC  ERROR no TLs are selected for CL_BASIC
[1656170916.449190] [aurora-r10:49843:0]         ucc_lib.c:127  UCC  ERROR lib_init failed for component: basic
[aurora-r10:49843:0:49843] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x164)
==== backtrace (tid:  49843) ====
 0  /home/niranda/miniconda3/envs/cylon_dev/lib/libucs.so.0(ucs_handle_error+0x2fd) [0x7f3eff20778d]
 1  /home/niranda/miniconda3/envs/cylon_dev/lib/libucs.so.0(+0x2b994) [0x7f3eff207994]
 2  /home/niranda/miniconda3/envs/cylon_dev/lib/libucs.so.0(+0x2bb5a) [0x7f3eff207b5a]
 3  /lib/x86_64-linux-gnu/libc.so.6(+0x43090) [0x7f3f00071090]
 4  /home/niranda/git/ucc/install/lib/libucc.so.1(ucc_collective_init+0x1e0) [0x7f3eff09bf30]
 5  /home/niranda/git/cylon/install/lib/libcylon.so.0.5.0(_ZNK5cylon3ucc21UccTableAllgatherImpl20AllgatherBufferSizesEPKiiPi+0x8e) [0x7f3f005f2b9e]
 6  /home/niranda/git/cylon/install/lib/libcylon.so.0.5.0(_ZN5cylon3net18TableAllgatherImpl7ExecuteERKSt10shared_ptrINS_15TableSerializerEERKS2_INS_9AllocatorEEiPSt6vectorIiSaIiEEPSB_IS2_INS_6BufferEESaISG_EEPSB_ISD_SaISD_EE+0xfc) [0x7f3f008d44dc]
 7  /home/niranda/git/cylon/install/lib/libcylon.so.0.5.0(_ZN5cylon3net18TableAllgatherImpl7ExecuteERKSt10shared_ptrINS_5TableEEPSt6vectorIS4_SaIS4_EE+0x1d1) [0x7f3f008d4ea1]
 8  /home/niranda/git/cylon/install/lib/libcylon.so.0.5.0(_ZNK5cylon3net15UCXCommunicator9AllGatherERKSt10shared_ptrINS_5TableEEPSt6vectorIS4_SaIS4_EE+0x5c) [0x7f3f005f192c]
 9  ./bin/ucc_allgather_example(+0x5f39) [0x55fbb51b8f39]
10  ./bin/ucc_allgather_example(+0x5c69) [0x55fbb51b8c69]
11  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7f3f00052083]
12  ./bin/ucc_allgather_example(+0x5e1e) [0x55fbb51b8e1e]
=================================
Segmentation fault (core dumped)

```

**Comment by kaiyingshan** [comment]
> 

I don't know if this is the same issue that I experienced, which is due to the ucc team size. When I run the example this way, it gives a warning "ucc_team.c:114  UCC  WARN  minimal size of UCC team is 2, provided 1", and it is caused by `ucc_collective_init`; when I run with `mpirun -n` it won't result in segfault.

File: cpp/src/cylon/net/ucx/ucx_communicator.cpp [hunk 1]
```diff
@@ -123,6 +146,56 @@ Status UCXCommunicator::Init(const std::shared_ptr<CommConfig> &config) {
...
+  ctx_params.oob.allgather = oob_allgather;
+  ctx_params.oob.req_test = oob_allgather_test;
+  ctx_params.oob.req_free = oob_allgather_free;
+  ctx_params.oob.coll_info = (void *)MPI_COMM_WORLD;
+  ctx_params.oob.n_oob_eps = static_cast<uint32_t>(this->world_size);
+  ctx_params.oob.oob_ep = static_cast<uint32_t>(rank);
+
+  RETURN_CYLON_STATUS_IF_UCC_FAILED(ucc_context_config_read(lib, nullptr, &ctx_config));
+  RETURN_CYLON_STATUS_IF_UCC_FAILED(ucc_context_create(lib, &ctx_params, ctx_config, &uccContext));
+  while (UCC_OK != ucc_context_progress(uccContext)) {
```
Comment (Sergei-Lebedev) [inline cpp/src/cylon/net/ucx/ucx_communicator.cpp:179]: ucc_context_create is blocking, no need to call ucc_context_progress

**Comment by Sergei-Lebedev** [comment]
> > 
> 
> I don't know if this is the same issue that I experienced, which is due to the ucc team size. When I run the example this way, it gives a warning "ucc_team.c:114 UCC WARN minimal size of UCC team is 2, provided 1", and it is caused by `ucc_collective_init`; when I run with `mpirun -n` it won't result in segfault.

recently we added support for team size 1 (https://github.com/openucx/ucc/pull/511)

**Comment by nirandaperera** [comment]
@kaiyingshan I reviewed your code and made some changes myself in this commit https://github.com/cylondata/cylon/pull/591/commits/9a7e9aa1ea21389963c6ac9d988f32381ab1b128

Could you please check that?

**Comment by nirandaperera** [comment]
@kaiyingshan thank you for doing this

**Pull request merged**